internal/controller/infrastructure/kwokmachine_controller.go

```diff
@@ -47,13 +55,122 @@ type KwokMachineReconciler struct {
 // For more details, check Reconcile and its Result here:
 // - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
 func (r *KwokMachineReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	_ = log.FromContext(ctx)
+	logger := ctrl.LoggerFrom(ctx)
 
-	// TODO(user): your logic here
+	kwokMachine := &infrastructurev1alpha1.KwokMachine{}
+	err := r.Get(ctx, req.NamespacedName, kwokMachine)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Fetch the machine
+	machine, err := util.GetOwnerMachine(ctx, r.Client, kwokMachine.ObjectMeta)
+	if err != nil {
+		logger.Error(err, "Failed to retrieve owner Machine from the API Server")
+
+		return ctrl.Result{}, err
+	}
+
+	if machine == nil {
+		logger.Info("Machine Controller has not yet set OwnerRef")
+
+		return ctrl.Result{Requeue: true}, nil
+	}
+
+	logger = logger.WithValues("machine", machine.Name)
+
+	if kwokMachine.Status.FailureMessage != nil ||
+		kwokMachine.Status.FailureReason != nil {
+		// TODO(vadasambar): add better log msg
+		// with more details (machine name, failure msg/reason etc.,)
+		logger.Info("KwokMachine has failed status")
+		return ctrl.Result{}, nil
+	}
+
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		logger.Info("Machine is missing cluster label or cluster does not exist")
+		return ctrl.Result{}, nil
+	}
+
+	if annotations.IsPaused(cluster, kwokMachine) {
+		logger.Info("Machine or linked Cluster is marked as paused. Won't reconcile")
+		return ctrl.Result{}, nil
+	}
+
+	logger = logger.WithValues("cluster", klog.KObj(cluster))
+
+	if cluster.Status.InfrastructureReady == false {
+		logger.Info("Cluster is not ready yet")
+		return ctrl.Result{}, nil
+	}
+
+	if machine.Spec.Bootstrap.DataSecretName == nil {
+		logger.Info("Bootstrap data secret is not present")
+		return ctrl.Result{}, nil
+	}
+
+	infraCluster, err := r.getInfraCluster(ctx, &logger, cluster, kwokMachine)
+
+	if err != nil {
+		return ctrl.Result{}, errors.New("error getting infra provider cluster or control plane object")
+	}
+	if infraCluster == nil {
+		logger.Info("KwokCluster or KwokManagedControlPlane is not ready yet")
+		return ctrl.Result{}, nil
+	}
+
+	if !machine.ObjectMeta.DeletionTimestamp.IsZero() {
+		err := r.Delete(context.Background(), kwokMachine)
+		if err != nil {
+			logger.Info("Error deleting KwokMachine", "kwokMachine", kwokMachine.GetName())
+			return ctrl.Result{Requeue: true}, nil
+		} else {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	// kwokMachine
```

Comment:
TODO: create `Node` here

Creating control plane machine using kwok might not be straightforward. kwok can create control plane but I am not sure if it can be represented as a Kubernetes `Node` resource. Another option here can be to create 1 node kind cluster and use it as a control plane `Machine` (this also creates dependency on the `kind` provider)

Reply:
![image](https://github.com/capi-samples/cluster-api-provider-kwok/assets/34534103/c3aac9c6-2b93-49ac-8198-ded69b531852)